.github/workflows/tests.yml

```diff
@@ -51,17 +51,27 @@ jobs:
       - name: Test with tox
         run: tox
 
-      - name: "Upload coverage to Codecov"
-        if: github.repository_owner == 'Uninett'
-        uses: codecov/codecov-action@v4
-        with:
-          fail_ci_if_error: true
-          token: ${{ secrets.CODECOV_TOKEN }}
-
       - name: Upload test reports (${{ matrix.python-version }})
         if: always()
-        uses: actions/upload-artifact@v3
+        uses: actions/upload-artifact@v4
         with:
-          name: reports
+          name: reports-${{ matrix.python-version }}
```

Comment:
I think that this step could potentially also be moved outside the matrix.  It seems the old version uploaded the same artifact several times, but the latest upload would include the test reports generated by all matrix dimensions (I assume the all work on the same file system) - so why would we need multiple artifacts if all the report files are available in the same directory structure after the matrix finishes?

Reply:
No, the old version also uploads an artifact for each Python version, which you can see for example in [this workflow](https://github.com/Uninett/zino/actions/runs/9268673231/job/25497813462) - the difference is just that they are merged into one folder
But yes, it is a good idea to move the uploading step outside to upload all test results at the same time

Reply:
I think you misunderstood me.  The old version clearly uploads multiple times, once for each Python version, which I wrote. But the artifact is named the same on every upload, so it's really the same artifact.

The question is whether the resulting artifact contains all test reports because of the old uploader action's magic artifact-merging functionality, or because all matrix dimensions share a common workspace directory?

If all matrix dimensions share a common workspace directory, the contents of the `report/` directory will grow for each matrix dimension that runs, meaning that each artifact step will just package more and more data, regardless of any magic artifact-merging.

If the matrix dimensions operate on isolated workspace directories, then we were dependent on the artifact-merging, and using the new uploader means we *do* have to produce multiple artifacts.  If the workspace is shared, we can get away with a single artifact produced at the end of the test run.

I genuinely don't know which is the case with workflow matrixes, and if you also don't, we should check the docs :)

Reply:
Ah, I see, yes, the old uploader is able to merge these artifacts. I read somewhere in my research, that previously these artifact were mutable, but with the new action version they are not anymore. 

I work under the assumption that all matrix dimensions share a common workspace directory, I am currently in the process of testing if that is the case and that we can that way upload the report artifact after all tests have run.

Reply:
From the release announcement of the artifacts action: "Using upload-artifact@v4 ensures artifacts are immutable"

Reply:
I suspected that could be the case.  But you didnt't change anything, and I didn't ask for this PR to change anything (it was a potential future issue to work on) - I approved the PR, so no need to re-request and wait for another review :smile: